apps/06SpaceTravelToMars.js

```diff
@@ -3,57 +3,138 @@
 var assert = require('assert');
 
 var jobTypes = {
-  pilot: 'MAV',
-  mechanic: 'Repair Ship',
-  commander: 'Main Ship',
-  programmer: 'Any Ship!'
+    pilot: 'MAV',
+    mechanic: 'Repair Ship',
+    commander: 'Main Ship',
+    programmer: 'Any Ship!'
 };
 
 // Your code here
 
+function CrewMember(name, job, specialSkill) {
+    this.name = name;
+    this.job = job;
+    this.specialSkill = specialSkill;
+    //this.ship = [];
+    this.ship = null;
+
+    this.enterShip = function(newShip) {
+
+        //assign the newShip to the crew member's ship
+        this.ship = newShip;
+
+        //push the crewmember into the ship's crew
+        newShip.crew.push(this);
+    }
+}
+
+function Ship(name, type, ability) {
+    this.name = name;
+    this.type = type;
+    this.ability = ability;
+    this.crew = [];
+
+    //console.log('crew length ' + this.crew.length);
+
+    this.missionStatement = function() {
+
+        // create a variable to count the valid jobs
+        var countValidJobs = 0;
+
+        // check if there are crew members assigned to this ship.
+        // if there are no crew members, return "can't perform a mission"
+        // if yes, then you loop through the crew members comparing their job to the jobtypes array to see if there is a match.  If there is a match, increase the countValidJobs by one.  If no match, return "can't perform a mission"
+
+        //if no crew members, can't perform a mission
+        if (this.crew.length === 0) {
+            return "Can't perform a mission yet.";
+        }
+
+        //if there are crew members, let's loop through to see if we have any with the correct jobs
+        else {
+
+            for (var i = 0; i < this.crew.length; i++) {
+
+                if (jobTypes[this.crew[i]['job']] == this.type) {
+                    countValidJobs++;
+                }
+            }
+
+            //Look at countValidJobs variable to see if there are any valid jobs. (should be greater than zero)
+            if (countValidJobs > 0) {
```

Comment:
I like this approach. If you can remember, find me in class and let's talk about if block nesting and 'early exit' scenarios